README.md
# KakaJSON
[![pod](https://img.shields.io/cocoapods/v/KakaJSON.svg)](https://github.com/CocoaPods/CocoaPods) [![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage) [![SwiftPM compatible](https://img.shields.io/badge/SwiftPM-Compatible-brightgreen.svg)](https://swift.org/package-manager/)

>  Fast conversion between JSON and model in Swift.

- Convert model to JSON with one line of code.（一行代码Model转JSON）
- Convert JSON to Model with one line of code.（一行代码JSON转Model）
- Archive\Unarchive object with one line of code.（一行代码实现常见数据的归档\解档）



## 中文教程
- [KakaJSON手册](https://www.cnblogs.com/mjios/p/11352776.html)



## Integration

### CocoaPods
```ruby
pod 'KakaJSON', '~> 1.1.2' 
```

### Carthage
```ruby
github "kakaopensource/KakaJSON" ~> 1.1.2
```

### Swift Package Manager

To use Swift Package Manager, you should update to Xcode 11.
* Open your project.
* Click File tab
* Select Swift Packages 
* Add Package Dependency, enter [KakaJSON repo's URL](https://github.com/kakaopensource/KakaJSON.git)

Or you can login Xcode with your GitHub account. just search **KakaJSON**.



## Usages
- [Coding](#coding)
- [JSON To Model_01_Basic Usage](#json-to-model_01_basic-usage)
  - [Simple Model](#simple-model)
  - [Class Type](#class-type)
  - [Inheritance](#inheritance)
  - [let](#let)
  - [NSNull](#nsnull)
  - [JSONString](#jsonstring)
  - [JSONData](#jsondata)
  - [Nested Model 1](#nested-model-1)
  - [Nested Model 2](#nested-model-2)
  - [Nested Model 3](#nested-model-3)
  - [Recursive](#recursive)
  - [Generic](#generic)
  - [Model Array](#model-array)
  - [Model Array In Dictionary](#model-array-in-dictionary)
  - [Convert](#convert)
  - [Listen](#listen)
- [JSON To Model_02_Data Type](#json-to-model_02_data-type)
  - [Int](#int)
  - [Float](#float)
  - [Double](#double)
  - [CGFloat](#cgfloat)
  - [Bool](#bool)
  - [String](#string)
  - [Decimal](#decimal)
  - [NSDecimalNumber](#nsdecimalnumber)
  - [NSNumber](#nsnumber)
  - [Optional](#optional)
  - [URL](#url)
  - [Data](#data)
  - [Date](#date)
  - [Enum](#enum)
  - [Enum In Array](#enum-in-array)
  - [Enum In Dictionary](#enum-in-dictionary)
  - [Enum Array In Dictionary](#enum-array-in-dictionary)
  - [Array](#array)
  - [Set](#set)
  - [Dictionary](#dictionary)
- [JSON To Model_03_Key Mapping](#json-to-model_03_key-mapping)
  - [Basic Usage](#basic-usage)
  - [Camel -> Underline](#camel---underline)
  - [Underline -> Camel](#underline---camel)
  - [Inheritance](#inheritance-1)
  - [Override 1](#override-1)
  - [Override 2](#override-2)
  - [Global Config](#global-config)
  - [Local Config](#local-config)
  - [Config Example 1](#config-example-1)
  - [Config Example 2](#config-example-2)
  - [Complex](#complex)
- [JSON To Model_04_Custom Value](#json-to-model_04_custom-value)
  - [Date](#date-1)
  - [Unspecific Type](#unspecific-type)
  - [Example](#example)
  - [Other Ways](#other-ways)
- [JSON To Model_05_Dynamic Model](#json-to-model_05_dynamic-model)
- [Model To JSON](#model-to-json)
	- [JSON and JSONString](#json-and-jsonstring)
	- [Optional](#optional-1)
	- [Enum](#enum-1)
	- [Nested Model](#nested-model)
	- [Any](#any)
	- [Model Array](#model-array-1)
	- [Model Set](#model-set)
	- [Key Mapping](#key-mapping-1)
	- [Custom Value](#custom-value-1)
	- [Listen](#listen)



## Coding

```swift
// file path (can be String or URL)
let file = "/Users/mj/Desktop/test.data"

/****************** String ******************/
let string1 = "123"
// wrtite String to file
write(string1, to: file)
// read String from file
let string2 = read(String.self, from: file)
XCTAssert(string2 == string1)
// read Int from file
XCTAssert(read(Int.self, from: file) == 123)

/****************** Date ******************/
let date1 = Date(timeIntervalSince1970: 1565922866)
// wrtite Date to file
write(date1, to: file)
 
// read Date from file
let date2 = read(Date.self, from: file)
XCTAssert(date2 == date1)
 
// read Int from file
XCTAssert(read(Int.self, from: file) == 1565922866)

/****************** Array ******************/
let array1 = ["Jack", "Rose"]
// wrtite [String] to file
write(array1, to: file)
 
// read [String] from file
let array2 = read([String].self, from: file)
XCTAssert(array2 == array1)
// Also support Set\Dictionary

/****************** Model ******************/
struct Book: Convertible {
    var name: String = ""
    var price: Double = 0.0
}
 
struct Car: Convertible {
    var name: String = ""
    var price: Double = 0.0
}
 
struct Dog: Convertible {
    var name: String = ""
    var age: Int = 0
}
 
struct Person: Convertible {
    var name: String = "Jack"
    var car: Car? = Car(name: "Bently", price: 106.666)
    var books: [Book]? = [
        Book(name: "Fast C++", price: 666.6),
        Book(name: "Data Structure And Algorithm", price: 666.6),
    ]
    var dogs: [String: Dog]? = [
        "dog0": Dog(name: "Wang", age: 5),
        "dog1": Dog(name: "ErHa", age: 3),
    ]
}
 
// wrtite Person to file
write(Person(), to: file)
 
// read Person from file
let person = read(Person.self, from: file)
 
XCTAssert(person?.name == "Jack")
XCTAssert(person?.car?.name == "Bently")
XCTAssert(person?.car?.price == 106.666)
XCTAssert(person?.books?.count == 2)
XCTAssert(person?.dogs?.count == 2)

/****************** Model Array ******************/
struct Car: Convertible {
    var name: String = ""
    var price: Double = 0.0
}
 
let models1 = [
    Car(name: "BMW", price: 100.0),
    Car(name: "Audi", price: 70.0)
]
// wrtite [Car] to file
write(models1, to: file)
 
// read [Car] from file
let models2 = read([Car].self, from: file)
XCTAssert(models2?.count == models1.count)
XCTAssert(models2?[0].name == "BMW")
XCTAssert(models2?[0].price == 100.0)
XCTAssert(models2?[1].name == "Audi")
XCTAssert(models2?[1].price == 70.0)

/****************** Model Set ******************/
struct Car: Convertible, Hashable {
    var name: String = ""
    var price: Double = 0.0
}
 
let models1: Set<Car> = [
    Car(name: "BMW", price: 100.0),
    Car(name: "Audi", price: 70.0)
]
 
// wrtite Set<Car> to file
write(models1, to: file)
 
// read Set<Car> from file
let models2 = read(Set<Car>.self, from: file)!
XCTAssert(models2.count == models1.count)
for car in models2 {
    XCTAssert(["BMW", "Audi"].contains(car.name))
    XCTAssert([100.0, 70.0].contains(car.price))
}

/****************** Model Dictionary ******************/
struct Car: Convertible {
    var name: String = ""
    var price: Double = 0.0
}
 
let models1 = [
    "car0": Car(name: "BMW", price: 100.0),
    "car1": Car(name: "Audi", price: 70.0)
]
 
// wrtite [String: Car] to file
write(models1, to: file)
 
// read [String: Car] from file
let models2 = read([String: Car].self, from: file)
XCTAssert(models2?.count == models1.count)
 
let car0 = models2?["car0"]
XCTAssert(car0?.name == "BMW")
XCTAssert(car0?.price == 100.0)
 
let car1 = models2?["car1"]
XCTAssert(car1?.name == "Audi")
XCTAssert(car1?.price == 70.0)
```



## JSON To Model_01_Basic Usage

### Simple Model
```swift
struct Cat: Convertible {
    var name: String = ""
    var weight: Double = 0.0
}

// json can also be NSDictionary, NSMutableDictionary
let json: [String: Any] = [
    "name": "Miaomiao",
    "weight": 6.66
]

let cat1 = json.kj.model(Cat.self)
XCTAssert(cat1.name == "Miaomiao")
XCTAssert(cat1.weight == 6.66)

// you can call global function `model`
let cat2 = model(from: json, Cat.self)

// support type variable
var type: Convertible.Type = Cat.self
let cat3 = json.kj.model(type: type) as? Cat
let cat4 = model(from: json, type: type) as? Cat
```

### Class Type
```swift
class Cat: Convertible {
    var weight: Double = 0.0
    var name: String = ""
    // The protocol `Convertible` required an init constructor
    // for initializing an instance completely.
    required init() {}
}
let json = ...
let cat = json.kj.model(Cat.self)

// a class inherit from NSObject
class Person: NSObject, Convertible {
    var name: String = ""
    var age: Int = 0
    // must add `override` because NSObject has `init`
    required override init() {}
}
let person = json.kj.model(Person.self)

struct Dog: Convertible {
    var weight: Double = 0.0
    var name: String = ""
    // This struct don't need to implement init because compiler generates init for it
}

struct Pig: Convertible {
    var weight: Double
    var name: String
    // This struct need to implement init to initialize all the stored properties.
    init() {
        name = ""
        weight = 0.0
    }
}
```

### Inheritance
```swift
class Person: Convertible {
    var name: String = ""
    var age: Int = 0
    required init() {}
}
 
class Student: Person {
    var score: Int = 0
    var no: String = ""
}
 
let json: [String: Any] = [
    "name": "jack",
    "age": 18,
    "score": 98,
    "no": "9527"
]
 
let student = json.kj.model(Student.self)
```

### let
```swift
struct Cat: Convertible {
    // let of integer type is very restricted in release mode
    // please user `private(set) var` instead of `let`
    private(set) var weight: Double = 0.0
    let name: String = ""
}
let json = ...
let cat = json.kj.model(Cat.self)
```

### NSNull
```swift
struct Cat: Convertible {
    var weight: Double = 0.0
    var name: String = "xx"
    var data: NSNull?
}

let json: [String: Any] = [
    "name": NSNull(),
    "weight": 6.6,
    "data": NSNull()
]

let cat = json.kj.model(Cat.self)
// convert failed, keep default value
XCTAssert(cat.name == "xx")
XCTAssert(cat.weight == 6.6)
XCTAssert(cat.data == NSNull())
```

### JSONString
```swift
// jsonString can alse be NSString, NSMutableString
let jsonString = """
{
    "name": "Miaomiao",
    "weight": 6.66
}
"""
 
let cat1 = jsonString.kj.model(Cat.self)
let cat2 = model(from: jsonString, Cat.self)

var type: Convertible.Type = Cat.self
let cat3 = jsonString.kj.model(type: type) as? Cat
let cat4 = model(from: jsonString, type: type) as? Cat
```

### JSONData
```swift
// jsonData can alse be NSData, NSMutableData
let jsonData = """
{
    "name": "Miaomiao",
    "weight": 6.66
}
""".data(using: .utf8)!
 
let cat1 = jsonData.kj.model(Cat.self)
let cat2 = model(from: jsonData, Cat.self)

var type: Convertible.Type = Cat.self
let cat3 = jsonData.kj.model(type: type) as? Cat
let cat4 = model(from: jsonData, type: type) as? Cat
```

### Nested Model 1
```swift
// let all the models comform to Convertible

struct Book: Convertible {
    var name: String = ""
    var price: Double = 0.0
}
 
struct Car: Convertible {
    var name: String = ""
    var price: Double = 0.0
}
 
struct Dog: Convertible {
    var name: String = ""
    var age: Int = 0
}
 
struct Person: Convertible {
    var name: String = ""
    var car: Car?
    var books: [Book]?
    var dogs: [String: Dog]?
}
 
let json: [String: Any] = [
    "name": "Jack",
    "car": ["name": "BMW7", "price": 105.5],
    "books": [
        ["name": "Fast C++", "price": 666.6],
        ["name": "Data Structure And Algorithm", "price": 1666.6]
    ],
    "dogs": [
        "dog0": ["name": "Larry", "age": 5],
        "dog1": ["name": "ErHa", "age": 2]
    ]
]
 
let person = json.kj.model(Person.self)
XCTAssert(person.car?.name == "BMW7")
XCTAssert(person.books?[1].name == "Data Structure And Algorithm")
XCTAssert(person.dogs?["dog0"]?.name == "Larry")
```

### Nested Model 2
```swift
struct Book: Convertible, Hashable {
    var name: String = ""
    var price: Double = 0.0
}
 
struct Person: Convertible {
    var name: String = ""
    var books: Set<Book>?
}
 
let json: [String: Any] = [
    "name": "Jack",
    "books": [
        ["name": "Fast C++", "price": 666.6]
    ]
]
 
let person = json.kj.model(Person.self)
XCTAssert(person.name == "Jack")
 
XCTAssert(person.books?.count == 1)
let book = person.books?.randomElement()
XCTAssert(book?.name == "Fast C++")
XCTAssert(book?.price == 666.6)
```

### Nested Model 3
```swift
struct Car: Convertible {
    var name: String = ""
    var price: Double = 0.0
}

class Dog: Convertible {
    var name: String = ""
    var age: Int = 0
    required init() {}
    init(name: String, age: Int) {
        self.name = name
        self.age = age
    }
}

struct Person: Convertible {
    var name: String = ""
    // KakaJSON will use your defaultValue instead of creating a new model
    // KakaJSON will not creat a new model again if you already have a default model value
    var car: Car = Car(name: "Bently", price: 106.5)
    var dog: Dog = Dog(name: "Larry", age: 5)
}

let json: [String: Any] = [
    "name": "Jake",
    "car": ["price": 305.6],
    "dog": ["name": "Wangwang"]
]

let person = json.kj.model(Person.self)
XCTAssert(person.name == "Jake")
// keep defaultValue
XCTAssert(person.car.name == "Bently")
// use value from json
XCTAssert(person.car.price == 305.6)
// use value from json
XCTAssert(person.dog.name == "Wangwang")
// keep defaultValue
XCTAssert(person.dog.age == 5)
```

### Recursive
```swift
class Person: Convertible {
    var name: String = ""
    var parent: Person?
    required init() {}
}
 
let json: [String: Any] = [
    "name": "Jack",
    "parent": ["name": "Jim"]
]
 
let person = json.kj.model(Person.self)
XCTAssert(person.name == "Jack")
XCTAssert(person.parent?.name == "Jim")
```

### Generic
```swift
struct NetResponse<Element>: Convertible {
    let data: Element? = nil
    let msg: String = ""
    private(set) var code: Int = 0
}
 
struct User: Convertible {
    let id: String = ""
    let nickName: String = ""
}
 
struct Goods: Convertible {
    private(set) var price: CGFloat = 0.0
    let name: String = ""
}
 
let json1 = """
{
    "data": {"nickName": "KaKa", "id": 213234234},
    "msg": "Success",
    "code" : 200
}
"""
let response1 = json1.kj.model(NetResponse<User>.self)
XCTAssert(response1?.msg == "Success")
XCTAssert(response1?.code == 200)
XCTAssert(response1?.data?.nickName == "KaKa")
XCTAssert(response1?.data?.id == "213234234")

let json2 = """
{
    "data": [
        {"price": "6199", "name": "iPhone XR"},
        {"price": "8199", "name": "iPhone XS"},
        {"price": "9099", "name": "iPhone Max"}
    ],
    "msg": "Success",
    "code" : 200
}
"""
let response2 = json2.kj.model(NetResponse<[Goods]>.self)
XCTAssert(response2?.msg == "Success")
XCTAssert(response2?.code == 200)
XCTAssert(response2?.data?.count == 3)
XCTAssert(response2?.data?[0].price == 6199)
XCTAssert(response2?.data?[0].name == "iPhone XR")
XCTAssert(response2?.data?[1].price == 8199)
XCTAssert(response2?.data?[1].name == "iPhone XS")
XCTAssert(response2?.data?[2].price == 9099)
XCTAssert(response2?.data?[2].name == "iPhone Max")
```

### Model Array
```swift
struct Car: Convertible {
    var name: String = ""
    var price: Double = 0.0
}
 
// json can also be NSArray, NSMutableArray
let json: [[String: Any]] = [
    ["name": "Benz", "price": 98.6],
    ["name": "Bently", "price": 305.7],
    ["name": "Audi", "price": 64.7]
]
 
 
let cars1 = json.kj.modelArray(Car.self)
XCTAssert(cars1[1].name == "Bently")
 
let cars2 = modelArray(from: json, Car.self)

var type: Convertible.Type = Car.self
let cars3 = json.kj.modelArray(type: type) as? [Car]
let cars4 = modelArray(from: json, type: type) as? [Car]
 
// jsonString -> Model Array
let jsonString = "...."
let cars5 = jsonString.kj.modelArray(Car.self)
let cars6 = modelArray(from: jsonString, Car.self)
let cars7 = jsonString.kj.modelArray(type: type) as? [Car]
let cars8 = modelArray(from: jsonString, type: type) as? [Car]
```

### Model Array In Dictionary
```swift
struct Book: Convertible {
    var name: String = ""
    var price: Double = 0.0
}

struct Person: Convertible {
    var name: String = ""
    var books: [String: [Book?]?]?
}

let name = "Jack"
let mobileBooks = [
    (name: "iOS", price: 10.5),
    (name: "Android", price: 8.5)
]
let serverBooks = [
    (name: "Java", price: 20.5),
    (name: "Go", price: 18.5)
]

let json: [String: Any] = [
    "name": name,
    "books": [
        "mobile": [
            ["name": mobileBooks[0].name, "price": mobileBooks[0].price],
            ["name": mobileBooks[1].name, "price": mobileBooks[1].price]
        ],
        "server": [
            ["name": serverBooks[0].name, "price": serverBooks[0].price],
            ["name": serverBooks[1].name, "price": serverBooks[1].price]
        ]
    ]
]

let person = json.kj.model(Person.self)
XCTAssert(person.name == name)
let books0 = person.books?["mobile"]
XCTAssert(books0??.count == mobileBooks.count)
for i in 0..<mobileBooks.count {
    XCTAssert(books0??[i]?.name == mobileBooks[i].name);
    XCTAssert(books0??[i]?.price == mobileBooks[i].price);
}
let books1 = person.books?["server"]
XCTAssert(books1??.count == serverBooks.count)
for i in 0..<serverBooks.count {
    XCTAssert(books1??[i]?.name == serverBooks[i].name);
    XCTAssert(books1??[i]?.price == serverBooks[i].price);
}
```

### Convert
```swift
struct Cat: Convertible {
    var name: String = ""
    var weight: Double = 0.0
}
 
let json: [String: Any] = [
    "name": "Miaomiao",
    "weight": 6.66
]
 
var cat = Cat()
// fill a cat instance with json
// .kj_m is a mutable version of .kj
cat.kj_m.convert(json)
XCTAssert(cat.name == "Miaomiao")
XCTAssert(cat.weight == 6.66)

```

### Listen
```swift
struct Car: Convertible {
    var name: String = ""
    var age: Int = 0
    
    // call when will begin to convert from json to model
    mutating func kj_willConvertToModel(from json: JSONObject) {
        print("Car - kj_willConvertToModel")
    }
    
    // call when did finish converting from json to model
    mutating func kj_didConvertToModel(from json: JSONObject) {
        print("Car - kj_didConvertToModel")
    }
}
 
let name = "Benz"
let age = 100
let car = ["name": name, "age": age].kj.model(Car.self)
// Car - kj_willConvertToModel
// Car - kj_didConvertToModel
XCTAssert(car.name == name)
XCTAssert(car.age == age)
 
/*************************************************************/
 
class Person: Convertible {
    var name: String = ""
    var age: Int = 0
    required init() {}
    
    func kj_willConvertToModel(from json: JSONObject) {
        print("Person - kj_willConvertToModel")
    }
    
    func kj_didConvertToModel(from json: JSONObject) {
        print("Person - kj_didConvertToModel")
    }
}
 
class Student: Person {
    var score: Int = 0
    
    override func kj_willConvertToModel(from json: JSONObject) {
        // call super's implementation if necessary
        super.kj_willConvertToModel(from: json)
        
        print("Student - kj_willConvertToModel")
    }
    
    override func kj_didConvertToModel(from json: JSONObject) {
        // call super's implementation if necessary
        super.kj_didConvertToModel(from: json)
        
        print("Student - kj_didConvertToModel")
    }
}
 
let name = "jack"
let age = 10
let score = 100
let student = ["name": name, "age": age, "score": score].kj.model(Student.self)
// Person - kj_willConvertToModel
// Student - kj_willConvertToModel
// Person - kj_didConvertToModel
// Student - kj_didConvertToModel
XCTAssert(student.name == name)
XCTAssert(student.age == age)
XCTAssert(student.score == score)

```



## JSON To Model_02_Data Type

### Int
```swift
struct Student: Convertible {
    var age1: Int8 = 6
    var age2: Int16 = 0
    var age3: Int32 = 0
    var age4: Int64 = 0
    var age5: UInt8 = 0
    var age6: UInt16 = 0
    var age7: UInt32 = 0
    var age8: UInt64 = 0
    var age9: UInt = 0
    var age10: Int = 0
    var age11: Int = 0
}
 
let json: [String: Any] = [
    "age1": "suan8fa8",
    "age2": "6suan8fa8",
    "age3": "6",
    "age4": 6.66,
    "age5": NSNumber(value: 6.66),
    "age6": Int32(6),
    "age7": true,
    "age8": "FALSE", 
    "age9": Decimal(6.66),
    "age10": NSDecimalNumber(value: 6.66)，
    "age11": Date(timeIntervalSince1970: 1565922866)
]
 
let student = json.kj.model(Student.self)
// convert failed，keep defualt value
XCTAssert(student.age1 == 6) 
XCTAssert(student.age2 == 6) 
XCTAssert(student.age3 == 6)
XCTAssert(student.age4 == 6)
XCTAssert(student.age5 == 6)
XCTAssert(student.age6 == 6)
// true is 1，false is 0
XCTAssert(student.age7 == 1) 
// "true"\"TRUE"\"YES"\"yes" is 1，"false"\"FALSE"\"NO"\"no" is 0
XCTAssert(student.age8 == 0) 
XCTAssert(student.age9 == 6)
XCTAssert(student.age10 == 6)
XCTAssert(student.age11 == 1565922866)
```

### Float
```swift
struct Student: Convertible {
    var height1: Float = 0.0
    var height2: Float = 0.0
    var height3: Float = 0.0
    var height4: Float = 0.0
    var height5: Float = 0.0
    var height6: Float = 0.0
    var height7: Float = 0.0
    var height8: Float = 0.0
    var height9: Float = 0.0
}
 
let json: [String: Any] = [
    "height1": "6.66suan8fa8",
    "height2": "0.12345678",
    "height3": NSDecimalNumber(string: "0.12345678"),
    "height4": Decimal(string: "0.12345678") as Any,
    "height5": 666,
    "height6": true,
    "height7": "NO",
    "height8": CGFloat(0.12345678),
    "height9": Date(timeIntervalSince1970: 1565922866)
]
 
let student = json.kj.model(Student.self)
XCTAssert(student.height1 == 6.66)
XCTAssert(student.height2 == 0.12345678)
XCTAssert(student.height3 == 0.12345678)
XCTAssert(student.height4 == 0.12345678)
XCTAssert(student.height5 == 666.0)
// true is 1.0，false is 0.0
XCTAssert(student.height6 == 1.0)
// "true"\"TRUE"\"YES"\"yes" is 1.0，"false"\"FALSE"\"NO"\"no" is 0.0
XCTAssert(student.height7 == 0.0)
XCTAssert(student.height8 == 0.12345678)
XCTAssert(student.height9 == 1565922866)
```

### Double
```swift
struct Student: Convertible {
    var height1: Double = 0.0
    var height2: Double = 0.0

CHANGELOG.md
# Change Log

---

## [1.1.2](https://github.com/kakaopensource/KakaJSON/releases/tag/1.1.2) (2019-12-23)
- Supporting convert from [String: [[String: Any]]] to [String: [Model]]
- Supporting 'enum array', 'enum dict'
- Closed issues
	- [嵌套解析出错](https://github.com/kakaopensource/KakaJSON/issues/39)

## [1.1.1](https://github.com/kakaopensource/KakaJSON/releases/tag/1.1.1) (2019-11-04)
- Adjust Lock
- Closed issues
	- [iOS 9 Crash](https://github.com/kakaopensource/KakaJSON/issues/31)
- Merged pull requests
	- [use isUppercase to check if the character is uppercase](https://github.com/kakaopensource/KakaJSON/pull/20)
	- [A key should be able to contain dot](https://github.com/kakaopensource/KakaJSON/pull/21)
       - [use enum as namespace, as Apple does in Combine](https://github.com/kakaopensource/KakaJSON/pull/23)
       - [return early if possible](https://github.com/kakaopensource/KakaJSON/pull/24)
       - [add test testComplex_JSONKeyWithDot_bestMatch](https://github.com/kakaopensource/KakaJSON/pull/25)
       - [[big pr]: refactor project structure](https://github.com/kakaopensource/KakaJSON/pull/27)
       
       

## [1.1.0](https://github.com/kakaopensource/KakaJSON/releases/tag/1.1.0) (2019-08-28)
- Remove locks in ConvertibleConfig
- Simplify apis
- Support carthage
- Keep default model value for optimization
- Merged pull requests
	- [simplify optional.kj_value](https://github.com/kakaopensource/KakaJSON/pull/16)
	- [replace symbol ~= with pattern keyword](https://github.com/kakaopensource/KakaJSON/pull/17)



## [1.0.0](https://github.com/kakaopensource/KakaJSON/releases/tag/1.0.0) (2019-08-23)

- First public release


KakaJSONTests/Common.swift
//
//  Common.swift
//  KakaJSONTests
//
//  Created by MJ Lee on 2019/8/11.
//  Copyright © 2019 MJ Lee. All rights reserved.
//

class Common: XCTestCase {
    struct Cat: Convertible {
        var age: Int = 0
        var name: String = ""
    }
    
    func testJSON_To_Model() {
        let name = "Miaomiao"
        let age = 26
        let json: [String: Any] = [
            "age": age,
            "name": name
        ]
        let jsonString = """
        {
            "name": "\(name)",
            "age": \(age)
        }
        """
        let jsonData = jsonString.data(using: .utf8)
        
        let cat1 = model(from: json, Cat.self)
        XCTAssert(cat1?.name == name)
        XCTAssert(cat1?.age == age)
        
        let cat2 = model(from: json, anyType: Cat.self) as? Cat
        XCTAssert(cat2?.name == name)
        XCTAssert(cat2?.age == age)
        
        let cat3 = model(from: jsonString, Cat.self)
        XCTAssert(cat3?.name == name)
        XCTAssert(cat3?.age == age)
        
        let cat4 = model(from: jsonString, anyType: Cat.self) as? Cat
        XCTAssert(cat4?.name == name)
        XCTAssert(cat4?.age == age)
        
        let cat5 = model(from: jsonData, Cat.self)
        XCTAssert(cat5?.name == name)
        XCTAssert(cat5?.age == age)
        
        let cat6 = model(from: jsonData, anyType: Cat.self) as? Cat
        XCTAssert(cat6?.name == name)
        XCTAssert(cat6?.age == age)
    }
    
    func testModel_To_JSON() {
        let cat = Cat(age: 26, name: "Miaomiao")
        
        let json = JSONObject(from: cat)
        XCTAssert(json?["name"] as? String == cat.name)
        XCTAssert(json?["age"] as? Int == cat.age)
        
        let jsonString = JSONString(from: cat)
        XCTAssert(jsonString?.contains("\"age\":\(cat.age)") == true)
        XCTAssert(jsonString?.contains("\"name\":\"\(cat.name)\"") == true)
    }
    
    func testJSONArray_To_ModelArray() {
        let name = "Miaomiao"
        let age = 26
        let json: [[String: Any]] = [
            ["age": age, "name": name],
            ["age": age, "name": name]
        ]
        let jsonString = """
        [
            {"name": "\(name)", "age": \(age)},
            {"name": "\(name)", "age": \(age)}
        ]
        """
        
        let cats1 = modelArray(from: json, Cat.self)
        XCTAssert(cats1?[0].name == name)
        XCTAssert(cats1?[0].age == age)
        XCTAssert(cats1?[1].name == name)
        XCTAssert(cats1?[1].age == age)
        
        let cats2 = modelArray(from: json, anyType: Cat.self) as? [Cat]
        XCTAssert(cats2?[0].name == name)
        XCTAssert(cats2?[0].age == age)
        XCTAssert(cats2?[1].name == name)
        XCTAssert(cats2?[1].age == age)
        
        let cats3 = modelArray(from: jsonString, Cat.self)
        XCTAssert(cats3?[0].name == name)
        XCTAssert(cats3?[0].age == age)
        XCTAssert(cats3?[1].name == name)
        XCTAssert(cats3?[1].age == age)
        
        let cats4 = modelArray(from: jsonString, anyType: Cat.self) as? [Cat]
        XCTAssert(cats4?[0].name == name)
        XCTAssert(cats4?[0].age == age)
        XCTAssert(cats4?[1].name == name)
        XCTAssert(cats4?[1].age == age)
    }
    
    func testModelArray_To_JSONArray() {
        let cat = Cat(age: 26, name: "Miaomiao")
        let cats = [cat, cat]
        
        let json = JSONObjectArray(from: cats)
        XCTAssert(json?[0]["name"] as? String == cat.name)
        XCTAssert(json?[0]["age"] as? Int == cat.age)
        XCTAssert(json?[1]["name"] as? String == cat.name)
        XCTAssert(json?[1]["age"] as? Int == cat.age)
        
        let jsonString = JSONString(from: cats)
        XCTAssert(jsonString?.contains("[") == true)
        XCTAssert(jsonString?.contains("},{") == true)
        XCTAssert(jsonString?.contains("]") == true)
        XCTAssert(jsonString?.contains("\"age\":\(cat.age)") == true)
        XCTAssert(jsonString?.contains("\"name\":\"\(cat.name)\"") == true)
    }
}

let timeIntevalInt: Int = 1565922866
let timeIntevalFloat = Float(timeIntevalInt)
let timeInteval = Double(timeIntevalInt)
let timeIntevalString = "\(timeIntevalInt)"
let time = Date(timeIntervalSince1970: timeInteval)
// 16 decimals
let longDoubleString = "0.1234567890123456"
let longDouble: Double = 0.1234567890123456
// 8 decimals
let longFloatString = "0.12345678"
let longFloat: Float = 0.12345678
// 39 decimals
let longDecimalString = "0.123456789012345678901234567890123456789"
var longDecimal = Decimal(string: longDecimalString)!
var longDecimalNumber = NSDecimalNumber(string: longDecimalString)

func checkModelToJSon<M: Equatable & Convertible>(_ type: M.Type) {
    // create model
    let model = type.init()
    // get JSONString from model
    let jsonString = model.kj.JSONString()
    
    // check JSON and JSONString
    let modelFromJsonString = jsonString?.kj.model(anyType: type) as? M
    XCTAssert(model == modelFromJsonString)
    
    // prevent from 66.6499999999999998
    XCTAssert(jsonString?.contains("99999") == false)
    // prevent from 66.6600000000000001
    XCTAssert(jsonString?.contains("00000") == false)
}

/// note
/*
protocol Runnable {}
class Dog: Runnable {}

print("--------section0--------")
print(Runnable.self is Runnable.Type) // false

print("--------section1--------")
let dog1: Dog = Dog()
print(dog1 as? Runnable) // success
print(dog1 is Runnable) // true
print(dog1 as? Dog) // success
print(dog1 is Dog) // true

print("--------section2--------")
let dog2 = Dog.self
print(dog2 as? Runnable.Type) // success
print(dog2 is Runnable.Type) // true
print(dog2 as? Dog.Type) // success
print(dog2 is Dog.Type) // true

print("--------section3--------")
let dog3: Dog??? = Dog()
print(dog3 as? Runnable) // success
print(dog3 is Runnable) // true
print(dog3 as? Dog) // success
print(dog3 is Dog) // true

print("--------section4--------")
let dog4: Dog??? = nil
print(dog4 as? Runnable) // nil
print(dog4 is Runnable) // false
print(dog4 as? Dog) // nil
print(dog4 is Dog) // false

print("--------section5--------")
let dog5 = Dog???.self
print(dog5 as? Runnable.Type) // nil
print(dog5 is Runnable.Type) // false
print(dog5 as? Dog.Type) // nil
print(dog5 is Dog.Type) // false

print("--------section6--------")
let dog6: Dog = Dog()
print(dog6 as? Runnable???) // success
print(dog6 is Runnable???) // false
print(dog6 as? Dog???) // success
print(dog6 is Dog???) // false

print("--------section7--------")
let dog7 = Dog.self
print(dog7 as? Runnable???.Type) // nil
print(dog7 is Runnable???.Type) // false
print(dog7 as? Dog???.Type) // nil
print(dog7 is Dog???.Type) // false
 */

/*
class Runnable {}
class Dog: Runnable {}

print("--------section0--------")
print(Runnable.self is Runnable.Type) // true

print("--------section1--------")
let dog1: Dog = Dog()
print(dog1 as? Runnable) // success
print(dog1 is Runnable) // true
print(dog1 as? Dog) // success
print(dog1 is Dog) // true

print("--------section2--------")
let dog2 = Dog.self
print(dog2 as? Runnable.Type) // success
print(dog2 is Runnable.Type) // true
print(dog2 as? Dog.Type) // success
print(dog2 is Dog.Type) // true

print("--------section3--------")
let dog3: Dog??? = Dog()
print(dog3 as? Runnable) // success
print(dog3 is Runnable) // true
print(dog3 as? Dog) // success
print(dog3 is Dog) // true

print("--------section4--------")
let dog4: Dog??? = nil
print(dog4 as? Runnable) // nil
print(dog4 is Runnable) // false
print(dog4 as? Dog) // nil
print(dog4 is Dog) // false

print("--------section5--------")
let dog5 = Dog???.self
print(dog5 as? Runnable.Type) // nil
print(dog5 is Runnable.Type) // false
print(dog5 as? Dog.Type) // nil
print(dog5 is Dog.Type) // false

print("--------section6--------")
let dog6: Dog = Dog()
print(dog6 as? Runnable???) // success
print(dog6 is Runnable???) // false
print(dog6 as? Dog???) // success
print(dog6 is Dog???) // false

print("--------section7--------")
let dog7 = Dog.self
print(dog7 as? Runnable???.Type) // nil
print(dog7 is Runnable???.Type) // false
print(dog7 as? Dog???.Type) // nil
print(dog7 is Dog???.Type) // false
 */

/*
let array: [Int?] = [1, nil, 2]
print(array as? [Any]) // success
print(array as? [Int]) // nil
 */
